### PR TITLE
[DEV-2603] kyc transfer v2

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations (on [DFX services](https://dev.services.dfx.swiss))
  - Login/logout
  - Buy/sell payment request
  - KYC page

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
- [ ] Add migration
- [ ] merge webhooks
- [ ] synchronize userData, wallet job

File: src/subdomains/generic/user/models/wallet/wallet.entity.ts [hunk 1]
```diff
@@ -11,6 +12,17 @@ export enum AmlRule {
...
+export interface WebhookConfig {
+  paymentWebhook: WebhookConfigOption;
+  kycWebhook: WebhookConfigOption;
+}
+
+export enum WebhookConfigOption {
+  TRUE = 'True',
+  FALSE = 'False',
+  CONSENT_ONLY = 'ConsentOnly',
+}
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/wallet/wallet.entity.ts:24]: Maybe we should also have the option `WalletOnly` 😉

File: src/subdomains/generic/user/services/webhook/webhook.service.ts [hunk 2]
```diff
@@ -14,100 +13,114 @@ import { WebhookRepository } from './webhook.repository';
...
+    // user data webhooks
+    const additionalClients = userData.kycClientList.filter((w) => !webhooks.some(({ wallet }) => wallet.id === w));
+    for (const walletId of additionalClients) {
+      const wallet = await this.walletService.getByIdOrName(walletId);
+      if (wallet?.isValidForWebhook(type, true)) webhooks.push({ type, data, reason, userData, wallet });
+    }
+
+    for (const client of webhooks) {
+      await this.createAndSendWebhook(client);
+    }
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook.service.ts:93]: Do we need to load the wallet of the user, in case it's not loaded?

File: src/subdomains/generic/user/models/user-data/user-data.entity.ts [hunk 3]
```diff
@@ -237,6 +237,9 @@ export class UserData extends IEntity {
   @Column({ length: 256, nullable: true })
   identDocumentType: string;
 
+  @Column({ length: 256, nullable: true })
+  kycClients: string; // semicolon separated wallet id's
```
Comment (davidleomay) [inline src/subdomains/generic/user/models/user-data/user-data.entity.ts:241]: KYC client names need to be added to the `KycInfoDto`

File: src/subdomains/generic/kyc/controllers/kyc.controller.ts [hunk 4]
```diff
@@ -101,6 +108,18 @@ export class KycController {
...
+  @UseGuards(AuthGuard(), new RoleGuard(UserRole.ACCOUNT))
+  async addKycClient(@GetJwt() jwt: JwtPayload, @Query('client') walletName: string) {
+    return this.kycService.addKycClient(jwt.account,walletName);
+  }
+
+  @Delete('transfer')
+  @UseGuards(AuthGuard(), new RoleGuard(UserRole.ACCOUNT))
+  async removeKycClient(@GetJwt() jwt: JwtPayload, @Query('client') walletName: string) {
+    return this.kycService.removeKycClient(jwt.account, walletName);
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/controllers/kyc.controller.ts:121]: Use the KYC code (`@Headers(CodeHeaderName) code: string`), no JWT authentication (see above)

File: src/subdomains/generic/user/services/webhook/webhook.entity.ts [hunk 5]
```diff
@@ -21,9 +23,15 @@ export class Webhook extends IEntity {
...
   // References
-  @ManyToOne(() => User, { nullable: false })
+  @ManyToOne(() => User, { nullable: true })
   user: User;
 
+  @ManyToOne(() => UserData, { nullable: true, eager: true })
+  userData: UserData;
+
+  @ManyToOne(() => Wallet, { nullable: true, eager: true })
+  wallet: Wallet;
```
[the remark this answers is not in the document]
Comment (Yannick1712) [reply src/subdomains/generic/user/services/webhook/webhook.entity.ts:33]: Yeah could do that, I thought maybe we remove the user relation after update and synchronization, that's why I don't have done it

File: src/subdomains/generic/user/services/webhook/dto/account-merge-webhook.dto.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+import { WebhookDto, WebhookType } from './webhook.dto';
+
+export class AccountMergeWebhookData {
+  @ApiProperty()
+  master: number;
+
+  @ApiProperty()
+  slave: number;
+}
+
```
[the remark this answers is not in the document]
Comment (Yannick1712) [reply src/subdomains/generic/user/services/webhook/dto/account-merge-webhook.dto.ts:11]: Yeah I thought master and slave are familiar words in development but we could also call them old and new

File: src/subdomains/generic/user/services/webhook/dto/webhook.dto.ts [hunk 7]
```diff
@@ -4,10 +4,14 @@ export enum WebhookType {
   PAYMENT = 'Payment',
   KYC_CHANGED = 'KycChanged',
   KYC_FAILED = 'KycFailed',
+  ACCOUNT_MERGE = 'AccountMerge',
 }
```
[the remark this answers is not in the document]
Comment (Yannick1712) [reply src/subdomains/generic/user/services/webhook/dto/webhook.dto.ts:8]: Hm yeah don't know if thats not a little bit confusing, have to think about it 🤔  because if we change anything else on account information that would also be `AccountChanged` and "merging" in this case is something different I would say to "changing"

File: src/subdomains/generic/user/services/webhook/webhook.entity.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (davidleomay) [reply src/subdomains/generic/user/services/webhook/webhook.entity.ts:33]: Ah ok

File: src/subdomains/generic/user/services/webhook/dto/webhook.dto.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (davidleomay) [reply src/subdomains/generic/user/services/webhook/dto/webhook.dto.ts:8]: Hm yes 🤔 I was thinking from the "slave" perspective. As a slave, the only change you see is a new account id 😄

File: src/subdomains/generic/user/services/webhook/dto/account-merge-webhook.dto.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (davidleomay) [reply src/subdomains/generic/user/services/webhook/dto/account-merge-webhook.dto.ts:11]: Slave doesn't make sense IMO since the slave doesn't have any functionality after the merge ... it's just gone 💥 maybe we should even not include the slave ID in the data and set something like this:
```ts
export class AccountMergeWebhookData {
  @ApiProperty()
  accountId: number;
}
```
Then we could name it `AccountChanged` and just add some further data to this DTO, if there is the need for it in the future.
WDYT?

[the remark this answers is not in the document]
Comment (Yannick1712) [reply src/subdomains/generic/user/services/webhook/dto/account-merge-webhook.dto.ts:11]: But then I don't know what the id was before the merge? 
That would be confusing too or?

File: src/subdomains/generic/user/models/user-data/user-data.entity.ts (hunk 3 above)
Comment (Yannick1712) [reply src/subdomains/generic/user/models/user-data/user-data.entity.ts:241]: Which dto do you mean exactly? Cannot find a KycInfoDto

Comment (Yannick1712) [reply src/subdomains/generic/user/models/user-data/user-data.entity.ts:241]: KycClient names? So we have to load every wallet to get the name?

File: src/subdomains/generic/user/services/webhook/webhook.service.ts (hunk 2 above)
Comment (Yannick1712) [reply src/subdomains/generic/user/services/webhook/webhook.service.ts:93]: Yeah I think so, or?

File: src/subdomains/generic/user/models/wallet/wallet.entity.ts [hunk 8]
```diff
@@ -43,4 +55,28 @@ export class Wallet extends IEntity {
...
+  }
+
+  isValidForWebhook(type: WebhookType, consented: boolean): boolean {
+    switch (type) {
+      case WebhookType.KYC_CHANGED:
+      case WebhookType.KYC_FAILED:
+        return (
+          this.webhookConfigObject.kycWebhook === WebhookConfigOption.TRUE ||
+          (this.webhookConfigObject.kycWebhook === WebhookConfigOption.CONSENT_ONLY && consented)
+        );
```
[the remark this answers is not in the document]
Comment (Yannick1712) [reply src/subdomains/generic/user/models/wallet/wallet.entity.ts]: In which form do you think?

File: src/subdomains/generic/user/models/user-data/user-data.entity.ts (hunk 3 above)
Comment (davidleomay) [reply src/subdomains/generic/user/models/user-data/user-data.entity.ts:241]: Haha, true, it's `KycLevelDto`.

Comment (davidleomay) [reply src/subdomains/generic/user/models/user-data/user-data.entity.ts:241]: Yes, I would set the names.

File: src/subdomains/generic/user/models/wallet/wallet.entity.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (davidleomay) [reply src/subdomains/generic/user/models/wallet/wallet.entity.ts]: Just pushed

File: src/subdomains/generic/user/services/webhook/webhook.service.ts (hunk 2 above)
Comment (davidleomay) [reply src/subdomains/generic/user/services/webhook/webhook.service.ts]: Just pushed

File: src/subdomains/generic/kyc/dto/mapper/kyc-info.mapper.ts [hunk 9]
```diff
@@ -27,6 +33,7 @@ export class KycInfoMapper {
       kycLevel: userData.kycLevelDisplay,
       tradingLimit: userData.tradingLimit,
       twoFactorEnabled: userData.totpSecret != null,
+      kycClients: kycClients?.map((kc) => kc.name),
       language: LanguageDtoMapper.entityToDto(userData.language),
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/dto/mapper/kyc-info.mapper.ts:37]: Add `?? []`?

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 10]
```diff
@@ -161,7 +162,9 @@ export class KycService {
     const user = await this.getUser(kycHash);
     await this.verifyUserDuplication(user);
 
-    return KycInfoMapper.toDto(user, false);
+    const kycClients = await this.getKycClients(user);
+
+    return KycInfoMapper.toDto(user, false, kycClients);
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:167]: Would be easier to put these two lines into a helper method 😉

File: src/subdomains/generic/kyc/services/kyc.service.ts [hunk 11]
```diff
@@ -519,6 +526,10 @@ export class KycService {
     await this.stepLogRepo.save(entity);
   }
 
+  private async getKycClients(userData: UserData): Promise<Wallet[]> {
+    return Promise.all(userData.kycClientList?.map((k) => this.walletService.getByIdOrName(k)) ?? []);
+  }
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/services/kyc.service.ts:531]: It's probably faster to fetch all wallets and do the filtering (with userData.kycClientList) in the mapper.

File: src/subdomains/generic/kyc/dto/mapper/kyc-info.mapper.ts [hunk 12]
```diff
@@ -14,25 +15,30 @@ import { KycLevelDto, KycSessionDto } from '../output/kyc-info.dto';
...
       kycSteps.find((s) => s.status === KycStepStatus.IN_PROGRESS) ??
       kycSteps.find((s) => s.status === KycStepStatus.FAILED);
 
+    const userKycClients = kycClients?.filter((kc) => userData.kycClientList.includes(kc.id));
+
     const dto: KycLevelDto | KycSessionDto = {
       kycLevel: userData.kycLevelDisplay,
       tradingLimit: userData.tradingLimit,
       twoFactorEnabled: userData.totpSecret != null,
+      kycClients: userKycClients?.map((kc) => kc.name) ?? [],
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/dto/mapper/kyc-info.mapper.ts:35]: `?` and fallback could probably be removed with the new implementation? Also above.

File: src/subdomains/generic/kyc/dto/mapper/kyc-info.mapper.ts [hunk 13]
```diff
@@ -14,25 +15,30 @@ import { KycLevelDto, KycSessionDto } from '../output/kyc-info.dto';
...
+    kycClients: Wallet[],
+    currentStep?: KycStep,
+  ): KycLevelDto | KycSessionDto {
     const kycSteps = KycInfoMapper.getUiSteps(userData);
     currentStep ??=
       kycSteps.find((s) => s.status === KycStepStatus.IN_PROGRESS) ??
       kycSteps.find((s) => s.status === KycStepStatus.FAILED);
 
+    const userKycClients = kycClients?.filter((kc) => userData.kycClientList.includes(kc.id));
+
```
Comment (davidleomay) [inline src/subdomains/generic/kyc/dto/mapper/kyc-info.mapper.ts:30]: `userData.kycClientList` can be undefined. Maybe we should add a fallback in the getter, would be safer.

File: src/subdomains/generic/user/services/webhook/webhook.service.ts [hunk 14]
```diff
@@ -14,100 +13,121 @@ import { WebhookRepository } from './webhook.repository';
...
+  private async createAndSendWebhook(dto: CreateWebhookInput): Promise<Webhook | undefined> {
+    const exists = await this.webhookRepo.exist({
       where: {
         data: dto.data,
         type: dto.type,
         reason: dto.reason,
-        user: { id: dto.user.id },
-        lastTryDate: IsNull(),
+        userData: { id: dto.userData.id },
+        wallet: { id: dto.wallet.id },
```
Comment (davidleomay) [inline src/subdomains/generic/user/services/webhook/webhook.service.ts:110]: Why did you remove the `lastTryDate`?

**Review by davidleomay — APPROVED** [review]

**Review by davidleomay — APPROVED** [review]
